src/types.rs

```diff
@@ -638,6 +642,93 @@ impl BytesChain {
     }
 }
 
+pub struct SemaphoreQueue {
+    wait_quota: AtomicIsize, // need to go below 0 when reapplying settings
+    semaphore: Arc<Semaphore>,
+}
+
+impl SemaphoreQueue {
+    pub fn new(queue_size: isize, permits: usize) -> Self {
+        Self {
+            wait_quota: AtomicIsize::new(queue_size),
+            semaphore: Arc::new(Semaphore::new(permits)),
+        }
+    }
+
+    pub fn acquire(&self) -> Option<SemaphoreQueueGuard<'_>> {
+        let mut quota = self.wait_quota.load(Ordering::Acquire);
+        loop {
+            if quota == 0 {
+                return None;
+            } else {
+                match self.wait_quota.compare_exchange(
+                    quota,
+                    quota - 1,
+                    Ordering::Release,
+                    Ordering::Relaxed,
+                ) {
+                    Ok(_) => break,
+                    Err(current) => {
+                        quota = current;
+                        continue;
+                    }
+                }
+            }
+        }
+        Some(SemaphoreQueueGuard { queue: self })
+    }
+
+    pub fn available_permits(&self) -> f64 {
+        self.semaphore.available_permits() as f64
+    }
+
+    pub async fn apply_limit(&self, old_limit: usize, new_limit: usize) {
+        if new_limit > old_limit {
+            self.semaphore.add_permits(new_limit - old_limit);
+        } else {
+            for _ in 0..old_limit - new_limit {
+                self.semaphore.acquire().await.forget();
+            }
+        }
+    }
+
+    pub fn apply_queue_size(&self, old_size: isize, new_size: isize) {
+        if new_size > old_size {
+            self.wait_quota
+                .fetch_add(new_size - old_size, Ordering::Relaxed);
+        } else {
+            // let the wait_quota go negative, once all the guards are awaited and resolved they
+            // will increment it back to positive values
+            self.wait_quota
+                .fetch_sub(old_size - new_size, Ordering::Relaxed);
+        }
+    }
+
+    fn remove_from_queue(&self) {
+        self.wait_quota.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+pub struct SemaphoreQueueGuard<'a> {
+    queue: &'a SemaphoreQueue,
+}
+
+impl<'a> Future for SemaphoreQueueGuard<'a> {
+    type Output = SemaphorePermit<'a>;
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        let inner = Pin::into_inner(self);
+        let permit_fut = inner.queue.semaphore.acquire();
+        tokio::pin!(permit_fut);
+        match permit_fut.poll(cx) {
+            task::Poll::Ready(v) => {
+                inner.queue.remove_from_queue();
+                task::Poll::Ready(v)
+            }
+            task::Poll::Pending => task::Poll::Pending,
+        }
+    }
```

Comment:
I understand your concerns, but I don't think that this is the case here. As the Waker (in Waiter) is [stored inside the semaphore](https://github.com/tokio-rs/tokio/blob/2afa276c8c31f7eeafa42f112a8b8507078ce15a/tokio/src/sync/batch_semaphore.rs#L32-L41) itself (which doesn't go anywhere), it doesn't get dropped after Acquire gets dropped. In fact after Acquire gets polled and cannot make progress (e.g. due to contention), it [enqueues the waiter on semaphore's wait list](https://github.com/tokio-rs/tokio/blob/2afa276c8c31f7eeafa42f112a8b8507078ce15a/tokio/src/sync/batch_semaphore.rs#L329-L351), so that the Waker is guaranteed to be woken as long as the semaphore is around.